src/anemoi/models/models/encoder_processor_decoder.py

```diff
@@ -202,32 +199,41 @@ def _run_mapper(
         )
 
     def forward(self, x: Tensor, model_comm_group: Optional[ProcessGroup] = None) -> Tensor:
-        batch_size = x.shape[0]
-        ensemble_size = x.shape[2]
+        batch_size, _, ensemble_size, *_ = x.shape
 
         # add data positional info (lat/lon)
-        x_data_latent = torch.cat(
-            (
-                einops.rearrange(x, "batch time ensemble grid vars -> (batch ensemble grid) (time vars)"),
-                self.trainable_data(self.latlons_data, batch_size=batch_size),
-            ),
-            dim=-1,  # feature dimension
-        )
+        x_data_latent = {}
+        for data in self._graph_input_meshes:
+            x_data_latent[data] = torch.cat(
+                (
+                    einops.rearrange(x, "batch time ensemble grid vars -> (batch ensemble grid) (time vars)"),
+                    getattr(self, f"trainable_{data}")(getattr(self, f"latlons_{data}"), batch_size=batch_size),
+                ),
+                dim=-1,  # feature dimension
+            )
 
         x_hidden_latent = self.trainable_hidden(self.latlons_hidden, batch_size=batch_size)
 
         # get shard shapes
-        shard_shapes_data = get_shape_shards(x_data_latent, 0, model_comm_group)
+        shard_shapes_data = {}
+        for data in self._graph_input_meshes:
+            shard_shapes_data[data] = get_shape_shards(x_data_latent[data], 0, model_comm_group)
         shard_shapes_hidden = get_shape_shards(x_hidden_latent, 0, model_comm_group)
 
-        # Run encoder
-        x_data_latent, x_latent = self._run_mapper(
-            self.encoder,
-            (x_data_latent, x_hidden_latent),
-            batch_size=batch_size,
-            shard_shapes=(shard_shapes_data, shard_shapes_hidden),
-            model_comm_group=model_comm_group,
-        )
+        # Run encoders
+        x_latents = []
```

Comment:
What's the performance impact of this?

Usually you'd open a tensor for this type of operation.